### DISCUSSION TITLE
In fast.farm,use simulink to control each turbine

### DISCUSSION THREAD

**Discussion by clun1** [OpenFAST/openfast#1557] [General]
i have a queastion,can we use simulink to control each turbine in fast.farm?

**Comment by clun1** [comment]
I have this problem. How can I solve it?
FAST_InitializeAll:FAST_Init:FAST_ReadPrimaryFile:1/VTK_fps is not an integer multiple of DT. FAST will
output VTK information at 14.286 fps, the closest rate possible.
FAST_InitializeAll:InflowWind_Init:TSFF_ValidateInput: Cannot find TurbSim full-field wind input file:
'D:\Desktop\fast.farm\openfast-3.4.1\openfast-3.4.1\reg_tests\r-test\glue-codes\fast-farm\TSinflow\Wind/90m_12mps_twr.bts'

**Reply by jjonkman** [reply]
Dear @clun1,

An interface has not yet been developed between FAST.Farm and MATLAB/Simulink, although this is listed on our desired future work: https://openfast.readthedocs.io/en/dev/source/user/fast.farm/FutureWork.html.

Does your TurbSim output file exist at the specified path?

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,

thanks a lot for your response.I did not specify a path to the output file.
My idea is to use each fan in fast.farm as a model, and then use simulink to independently control it. Is this method feasible? If not, is there a way to control the fast.farm?

Best regards,

**Reply by jjonkman** [reply]
Dear @clun1,

The TurbSim output file is referenced from the InflowWind input file you are using within FAST.Farm

As I said, an interface has not yet been developed between FAST.Farm and MATLAB/Simulink, so, you can't do as you propose without changing the source code to develop this interface.  You can control each individual wind turbine in the farm with any of the available OpenFAST control options except for the Simulink or Labview interfaces; i.e., built-in control options, user-defined routines, or the Bladed-style dynamic library .  If you also intend to implement a wind farm super controller, this is only feasible right now when making use of the user-defined via Bladed-style dynamic library.

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,
Thanks for your reply.If l use user_defined routines to modify control,how do i get start?Is there any usage documentation?Where is the file in the picture below?
Best regards,
<img width="876" alt="image" src="https://user-images.githubusercontent.com/93849049/236763321-60f3b28c-d2f2-4cfa-a7ad-986d67973e3a.png">

**Reply by jjonkman** [reply]
Dear @clun1,

There are placeholders for each user-define routine in the OpenFAST source code.  For example, a dummy version of `PitchCntrl()` is in _UserSubs.f90_ and a full example is in _PitchCntrl_ACH.f90_ of the ServoDyn source code.  To make use of these user-defined routines, you'll need to update the source code within these routines and recompile OpenFAST; documentation on compiling OpenFAST is available here: https://openfast.readthedocs.io/en/dev/source/install/index.html#compile-from-source.

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,
Thank you for your patient instruction and response. I checked and found the following documents. Are these documents related to what we talked about before.And then when I open up these files it looks like this,
Best regards,

<img width="556" alt="image" src="https://user-images.githubusercontent.com/93849049/236856074-90cb3cac-ae34-4ef0-baba-8b50646ed705.png">
<img width="839" alt="image" src="https://user-images.githubusercontent.com/93849049/236856340-a73e3417-6568-4215-8fab-19142074a3b9.png">

**Reply by jjonkman** [reply]
Dear @clun1,

You are looking at compiled files.  The files you want to edit are the source (_.f90_) files, available here: https://github.com/OpenFAST/openfast/tree/main/modules/servodyn/src.

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,
Thank you for your patient guidance. I have another question. If I change the mode in the file to 3(the picture is as follows,it is NRELOffshrBsline5MW_Onshore_ServoDyn_WT1.dat), I would like to modify the the yaw Angle of wind turbine, how can I find the corresponding code file, and then go in to modify it
Best regards,
<img width="1164" alt="1683728451968" src="https://github.com/OpenFAST/openfast/assets/93849049/1a95eca8-68be-4d47-910a-98a2a31f2553">

**Reply by clun1** [reply]
Dear jjonkman,
I found this function in the document, whether there is a relationship between them?
Best regards,
<img width="949" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/09e170e3-d55a-42fc-a210-deedcbe2b571">

**Reply by jjonkman** [reply]
Dear @clun1,

Yes, the `UserYawCont()` routine that you found in _UserSubs.f90_ is the routine you can modify to implement your own yaw control logic.  Just like with routine `PitchCntrl()`, to make use of routine `UserYawCont()`, you'll need to update the source code within this routine and recompile OpenFAST

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,
I wondered what changes I should make if I wanted the wind to yaw at 30 degrees, and I tried many times without the yaw changing accordingly
Best regards,

**Reply by jjonkman** [reply]
Dear @clun1,

To set you fixed but nonzero nacelle-yaw angle you can set the initial condition in ElastoDyn (`NacYaw`) with the nacelle-yaw DOF disabled (`YawDOF` = False).

If you want to use nacelle-yaw functionality of ServoDyn, you must have `YawDOF` = True in ElastoDyn.  The simple override maneuvers of ServoDyn permit you to define simple nacelle-yaw ramps.  More complicated nacelle-yaw control requires use of the one of the `YCMode` options.  The yaw commands from ServoDyn are passed through the nacelle-yaw actuator model as defined by ServoDyn inputs `YawSpr` and `YawDamp`.

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,
I want to use Super Controller(SC) to control each turbine in fast.farm, and i set UseSC ture ,but I don't know where to define the parameters of the Super Controller,Is there a manual? Where is the controller file?
Best regards,
<img width="1169" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/e07a0b62-317e-4e03-87af-ae2f88262074">
<img width="761" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/8bb1a17a-81ef-4c96-bcc6-394481b30f80">

**Reply by jjonkman** [reply]
Dear @clun1,

The use of the super controller in FAST.Farm is documented, e.g., see: https://openfast.readthedocs.io/en/main/source/user/fast.farm/ModelGuidance.html#super-controller.  And an example super controller (sc_dll, DISCON_SC) has been provided in my post dated March 18, 2021 in the following topic on our forum: https://forums.nrel.gov/t/fast-farm/1887.

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,
Thank you very much for your patient guidance and help me a lot. Thank you
Best regards,

**Reply by clun1** [reply]
Dear jjonkman,
Which file is the function called by DISCON.dll in Fast.farm? How can I know which one of the wind turbine controllers corresponds to? If I choose PCMode 5, what are the functions of these parameters in the servodyn file?
Best regards,
<img width="1112" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/59112ea1-2e37-48a4-a82d-6dc48c708aca">

**Reply by jjonkman** [reply]
Dear @clun1,

I'm not sure I understand your first two questions; please clarify.

Regarding the override pitch maneuvers, these will override the pitch commands coming from any of the pitch control options, including `PCMode` = 5.  E.g., if you have active control through a Bladed-style dynamic library starting at `TPCOn` = 0 s and `TPitManS(1)` = 100 s, then the dynamic library will control the pitch of blade 1 until 100 s into the simulation, after which `PitManRat(1)` and `BlPitchF(1)` will determine the pitch response of blade 1.

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,

Thank you for your patience.I want to use Bladed-style dynamic library(discon.dll) to contral the turbine of the fast.farm. But I don't know how to run successfully.Here's what I get when I run it.

Best regards,
<img width="939" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/b24a9089-e7c7-4e9a-ba6a-a575ddd7494f">

**Reply by jjonkman** [reply]
Dear @clun1,

Presumably you are using the FAST.Farm TSinflow example from the r-test: https://github.com/OpenFAST/r-test/tree/main/glue-codes/fast-farm/TSinflow.  Did you change this model in any way?   This model requires two copies of the DISCON controller dynamic library for the baseline NREL 5-MW turbine.  Where did you get these libraries?  Have you enabled the super controller within FAST.Farm (`UseSC` = TRUE)?

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,

thank for your reply,I already enabled the super controller within FAST.Farm(UseSC=TRUE), and add two copies of the DISCON controller dynamic library for the baseline NREL 5-MW turbine,here is the reasult as follow. Why is it happen? What parameters should I change?

Best regards,
<img width="1139" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/46cd8a65-075c-4dd2-8115-6e06006bf951">
<img width="394" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/552eff03-185e-41b7-97ce-2032807cb659">
<img width="849" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/f5130158-0d49-48b1-913d-c8e72de4e761">

**Reply by jjonkman** [reply]
Dear @clun1,

When the super controller is enabled in FAST.Farm, the interface to the individual controller dynamic library changes, requiring new input and output arguments to the dynamic library, as documented in the FAST.Farm modeling guidance here: https://openfast.readthedocs.io/en/main/source/user/fast.farm/ModelGuidance.html#super-controller.  My guess is you are not using the individual controller dynamic library with the interface that supports the super controller.  A similar question was asked and answered here: https://github.com/OpenFAST/openfast/discussions/1511.

Best regards,

**Comment by clun1** [comment]
Dear jjonkman,
 
I set all the parameters the same as in the forum, but there is still a problem. The results are as follows,how should I modify?
<img width="731" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/e4ebcca3-5102-4289-9488-809e4c47f52b">
<img width="721" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/d33747d0-f176-4cdb-95f0-5dda8294c54d">
<img width="706" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/f7169022-3d23-47e0-9ce4-ed3c90520597">
Best regards,

**Reply by jjonkman** [reply]
Dear @clun1,

Something is set improperly in your FAST.Farm or OpenFAST model, resulting in the model going numerically unstable.  But it is unclear to me from the errors reported what the problem is.  You mention that you've set, "all the parameters the same as in the forum".  I'm not exactly sure what this means.  I would suggest double checking your inputs and/or simplifying the model (e.g., disabling DOFs in ElastoDyn) to debug the issue.

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,

I have been wondering what is the relationship between Discon.dll and SC_dll.dll? Will they interact with data? How to use super controller to optimize the whole field?

Best regards,

**Reply by jjonkman** [reply]
Dear @clun1,

The super controller and its interaction with the individual turbine controllers are documented in the FAST.Farm modeling guidance; see: https://openfast.readthedocs.io/en/main/source/user/fast.farm/ModelGuidance.html#super-controller.

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,

1.When I use my own compiled fast.farm program(FAST.Farm_x64.exe), and then run fast.farm in cmd, the following situation occurs. How do I solve this problem?
<img width="844" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/4d3c3562-a48e-4c0c-bb9d-9620444a73ce">

2.If I want to make changes to supercontroller (SC_DLL) and Discon.dll and then compile, do I need to recompile fast.farm as well? Or I can use fast.farm_x64.exe，which I compiled earlier?I would like to know if there is a file describing the compilation relationship of these files?(except the openfast.readthedocs.io document)
<img width="182" alt="1688297539283" src="https://github.com/OpenFAST/openfast/assets/93849049/377a13f2-76a4-4830-b9c4-241b673c7ba0">

![2832343aafb238f5a88ef657457b471](https://github.com/OpenFAST/openfast/assets/93849049/e24b2674-0cc0-4724-aade-1a0f6b00cc04)

**Reply by jjonkman** [reply]
Dear @clun1,

Regarding (1), it sounds like your input file(s) are not compatible with the version of FAST.Farm you have compiled and are running.  InflowWind input parameter `VelInterpCubic` was added in OpenFAST v3.5.  The change in input file formatting with each release of OpenFAST is documented here: https://openfast.readthedocs.io/en/main/source/user/api_change.html.  And you can always enable the `Echo` option to debug errors in input file processing.

Regarding (2), changing and recompiling the controller dynamic libraries does not require you to recompile the OpenFAST or FAST.Farm executables, unless you've customized the interface in some way (added or removing arguments to the routines).

Best regards,

**Reply by clun1** [reply]
Dear jjonkman,

Thank you for your patient guidance,it really help me a lot,thank you.
1.Recently,I downloaded the new zip , set  the useSC true,and the following problem occurred(picture 1). When I set the useSC false,It works fine(picture 2).What is the cause of this problem and how can I solve it?
![image](https://github.com/OpenFAST/openfast/assets/93849049/26e84de2-b188-4d59-8f0f-87a88e3a9dd8)
<img width="689" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/1a1ac898-d2d3-4cec-aa0b-81391082f106">
<img width="852" alt="image" src="https://github.com/OpenFAST/openfast/assets/93849049/af8b3e92-c6ac-4e0a-8dc9-9388fea1948a">

2.Does the Discon.dll for fast.farm have to be compiled one at a time, and then copied to the path of fast.farm, From path D:\Desktop\new_fast\openfast-main \openfast-main\reg_tests\r-test\glue-codes\openfast\5MW_Baseline\ServoData Copy to D:\Desktop\new_fast\openfast-main \openfast-main\reg_tests\r-test\glue-codes\fast-farm\5MW_Baseline\ServoData.Is there a problem with this operation? Or does fast.farm have a one-time way of generating discon corresponding to the number of turbines in the wind farm? Have I described the problem clearly? I look forward to your guidance

Best regards,

**Reply by jjonkman** [reply]
Dear @clun1,

Regarding (1), I'm not sure.  I don't see any error messages written.  I would check that the name and path to your super controller is set properly in FAST.Farm input parameter `SC_FileName`.

Regarding (2), I'm not sure I fully understand your question.  You don't need to place the individual wind turbine controller(s) or super controller in a specific directory.  Just make sure that the file name and absolute or relative path are set correctly in FAST.Farm input parameter `SC_FileName` for the super controller, and ServoDyn input parameter `DLL_FileName` for the individual wind turbine controller(s).

Best regards,
